src/devices/src/pseudo/init.c

```diff
@@ -0,0 +1,45 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Init wrapper for boot timing.  Alpine-specific because it points at /sbin/openrc-init
```

Comment:
Why is this in `src/devices/` if it is Alpine specific?

Reply:
What Andreea said. We need to find a better location for this. Maybe `/tools`? @andreeaflorescu what do you think?

TLDR this file represents source of binary that writes to magic port on userspace init to trigger boot time calculation.

Reply:
Also, where is / when was this file used?

Reply:
> Why is this in `src/devices/` if it is Alpine specific?

Think I originally had it building via a `build.rs` script in the root of the devices module.  That bit got dropped but I left it where it is because there is obviously the connection between the signal and the boot_timer device handling it.  It made sense to me,  but happy to put it wherever is considered sensible. 

> Also, where is / when was this file used?

It lives in the guests that are used to measure boot time, and is a drop in wrapper replacement for `/sbin/init` on Alpine guests (a small tweak is required for systemd guests).  It's hopefully pretty close to what is currently on these guests.   So on a clean Alpine image `/sbin/int` is a symlink to `/sbin/openrc-init`,  whereas on the guests used to measure boot time, there is a wrapper program that sends the boot time signal before executing `/sbin/openrc-init`.

To see the current version in action,  boot into the `boottime-rootfs.ext4` rootfs,  tail the FC logs, and execute `/sbin/init` a few times.   You should hopefully see the boot time logging message being triggered on each execution.  To test this PR,  build Firecracker,  compile and copy over the "new" `/sbin/init` wrapper to the rootfs and repeat the `/sbin/init` calls.  Boot time logging should hopefully still be working.

Reply:
I don't think I quite grasp what is the purpose of this file. From what I understand, this file needs to be baked in the rootfs so that testing the bootime works. It seems to me that that this file is used for showcasing what needs to happen on `init`. Potentially you can build your own rootfs and include this file and then check the boot time. Are my assumptions correct?

Looking at this from a different angle:
- everything that is in `src/` is essentially shipped with the Firecracker release binary
- this file doesn't seem to be compiled
- is this file needed for release? if not, shouldn't it be somewhere in resources/tests?

Reply:
> Are my assumptions correct?

Yeah,  pretty much.  It's not just showcasing though, because it is needed for the bootime test.   Before merging this PR someone on the project side will need to compile this file and update the bootime rootfs images with the new init wrapper,  to correspond with the changes in this PR.   The main change being the address that the signal goes to i.e. currently on x86_64:  
```
const MAGIC_IOPORT_SIGNAL_GUEST_BOOT_COMPLETE: u64 = 0x03f0;
```
Which will no longer work when this change goes in,  because Firecracker will be listening for a signal at: 
```
0xd0000000
```
Which is the first slot on the MMIO bus,  i.e. the location of the BootTimer device.

`resources/tests`  sounds good to me,  I'll move it.

Reply:
@tnotw Thanks for the explanation, my question was more of a: why is the CI passing? :)) Was the rootfs image used by the boottime integration tests already updated in S3 with your init wrapper? :-? I think I'm missing something here.

Reply:
Ah, well,  I'm not sure it is to be honest.  I cannot see any `x86_64` logs being produced in the CI runs,  so I suspect something is rather broken.  And I have a suspicion that that boot time test for `aarch64` is disabled. 

To be honest,  I was still working on it and I'm not really sure it was 100% ready for review.

Reply:
Yep, it looks like the CI is not actually running if you open the PR as draft PR, but it's green and in the meantime it was marked as ready for review which made things even more confusing :(.
Also, boot time test for `aarch64` is not disabled.

Reply:
Ok,  so I'll try and trigger another run tonight and see what happens.

My other theory about why `'aarch64` works is that i've miscalculated where the bus timer lives on the MMIO bus,  and that both old and new locations are the same.  So it works with the old `/sbin/init` which points to `0x40000000` not  `0x40001000`, which I have in the new wrapper.   I've not done erm....... quite as much testing on `aarch64` as I would have liked :-)

Reply:
Seems to be better now,  in the sense that the right tests are failing,  i.e. the x86_64 boottime tests.